coramin/third_party/minlplib_tools.py

```diff
@@ -0,0 +1,336 @@
+        acceptable_formats = set([acceptable_formats])
+    elif isinstance(acceptable_formats, Iterable):
+        acceptable_formats = set(acceptable_formats)
+    else:
+        raise ValueError('unrecognized type for acceptable_formats: {0}'.format(acceptable_formats))
+        
+
+    if acceptable_probtype is None:
+        acceptable_probtype = set()
+        for pre in ['B', 'I', 'MI', 'MB', 'S', '']:
+            for post in ['NLP', 'QCQP', 'QP', 'QCP', 'P']:
+                acceptable_probtype.add(pre + post)
+    elif isinstance(acceptable_probtype, str):
+        acceptable_probtype = set([acceptable_probtype])
+    elif isinstance(acceptable_probtype, Iterable):
+        acceptable_probtype = set(acceptable_probtype)
+    else:
+        raise ValueError('unrecognized type for acceptable_probtype: {0}'.format(acceptable_probtype))
+
+    if acceptable_objtype is None:
+        acceptable_objtype = set(['constant', 'linear', 'quadratic', 'polynomial', 'signomial', 'nonlinear'])
+    elif isinstance(acceptable_objtype, str):
+        acceptable_objtype = set([acceptable_objtype])
+    elif isinstance(acceptable_objtype, Iterable):
+        acceptable_objtype = set(acceptable_objtype)
+    else:
+        raise ValueError('unrecognized type for acceptable_objtype: {0}'.format(acceptable_objtype))
+
+    if acceptable_objcurvature is None:
+        acceptable_objcurvature = set(['linear', 'convex', 'concave', 'indefinite', 'nonconvex', 'nonconcave', 'unknown'])
+    elif isinstance(acceptable_objcurvature, str):
+        acceptable_objcurvature = set([acceptable_objcurvature])
+    elif isinstance(acceptable_objcurvature, Iterable):
+        acceptable_objcurvature = set(acceptable_objcurvature)
+    else:
+        raise ValueError('unrecognized type for acceptable_objcurvature: {0}'.format(acceptable_objcurvature))
+
+    if acceptable_conscurvature is None:
+        acceptable_conscurvature = set(['linear', 'convex', 'concave', 'indefinite', 'nonconvex', 'nonconcave', 'unknown'])
+    elif isinstance(acceptable_conscurvature, str):
+        acceptable_conscurvature = set([acceptable_conscurvature])
+    elif isinstance(acceptable_conscurvature, Iterable):
+        acceptable_conscurvature = set(acceptable_conscurvature)
+    else:
+        raise ValueError('unrecognized type for acceptable_conscurvature: {0}'.format(acceptable_conscurvature))
+
+    if acceptable_convex is None:
+        acceptable_convex = set(['True', 'False', ''])
+    elif isinstance(acceptable_convex, str):
+        acceptable_convex = set([acceptable_convex])
+    elif isinstance(acceptable_convex, Iterable):
+        acceptable_convex = set(str(i) for i in acceptable_convex)
+    else:
+        acceptable_convex = set([str(acceptable_convex)])
+
+    csv_file = open(filename, 'r')
+    reader = csv.reader(csv_file, delimiter=';')
+    headings = {column: ndx for ndx, column in enumerate(next(reader))}
+    rows = [row for row in reader]
+    csv_file.close()
```

Comment:
Recommend (to guarantee file gets closed in the case of an exception reading the file):
```python
with open(filename, 'r') as csv_file:
    reader = csv.reader(csv_file, delimiter=';')
    headings = {column: ndx for ndx, column in enumerate(next(reader))}
    rows = [row for row in reader]
```

Reply:
Again, great suggestion. This is done as well.